### PR TITLE
Protonuke: Add FTP support

### PR DESCRIPTION
**Pull request by csymonds**
Adding in FTP traffic generation to protonuke
Supports TLS via -ftps

Client will connect and authenticate to a server, randomly choosing one of 6 ftp actions (including logging out) with uniform distribution.

File retrieval (RETR) is accomplished by sending an image file of specified size (default 500KB) over a passive data connection. The image is generated in the same way as http.go.

2 vendor packages were added to handle server and client implementation (with some additional implementation and updates required)

### CODE REVIEW

File: src/protonuke/protonuke.go [hunk 1]
```diff
@@ -15,34 +15,39 @@ import (
...
+	f_dnsv6           = flag.Bool("dnsv6", false, "dns client only requests type AAAA records")
+	f_randomhosts     = flag.Bool("random-hosts", false, "if no host range is supplied return a randomly generated ip")
+	f_http            = flag.Bool("http", false, "enable http service")
+	f_https           = flag.Bool("https", false, "enable https (TLS) service")
+	f_httproot        = flag.String("httproot", "", "serve directory with http(s) instead of the builtin page generator")
+	f_httpGzip        = flag.Bool("httpgzip", false, "gzip image served in http/https pages")
+	f_httpCookies     = flag.Bool("httpcookies", false, "enable cookie jar in http/https clients")
+	f_ftp             = flag.Bool("ftp", false, "enable ftp service")
+	f_ftps            = flag.Bool("ftps", false, "enable ftp (TLS) service")
+	f_ftpport         = flag.Int("ftpport", 21, "port to use for ftp client and server")
```
Comment (jcrussell) [inline src/protonuke/protonuke.go]: We don't have port flags for the other protocols, is there a reason to support ftp on a non-standard port?

File: src/protonuke/ftp.go [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
...elUCALTskjFd+ZsIfAslEgcDB22O+MwWcUTJvx/1OAswtXYJZ1HJ6xyZRBMf175USZjoHPhWd/g+BZRPpaAb6dqh/UYLrkI+OuUCD3WoUYt14g8y0geFQx8HvTZ+Hlv6S4dOIcFFD1Ipvnv8VkCQXf54e3HvqcixsqdxYR5v3hrTCOVu32IPysfTE+MIAtjvSJpTc/5vkOcVqmhEw+g/tGITTCqyWV2sIgJsfEvaCtpW/OjbAwrKFzngC2xeUG1wT737K5afT7kiENF9yz4cbztbrAxW/wa9sSqhHvqmWrotGC0ozvIVDtecSLmmuWCtu451ZM1iyW/7DUJe2FTSgQIDAQABo1MwUTAdBgNVHQ4EFgQUaiJnMwXWfzG8dQrI738tIIhECBQwHwYDVR0jBBgwFoAUaiJnMwXWfzG8dQrI738tIIhECBQwDwYDVR0TAQH/BAUwAwEB/zANBgkqhkiG9w0BAQ0FAAOCAgEAr/n81IPWeyD9X0UcheAnVPJI5oMyUA0lPF9nfVcyjuy0tI8x0GYM91B+5/CTHiy//724ZgK/cLN63O1H+cm6C9x1VlDuuIxnt+ht9+Saf1B0CO65zZKzi4k9uuWkmE50Fsa7cyoL3XyhZhnj1LtCWBW1s3Z5MPPH3ouYv/NSRfOC8bDoPGMGBzL4EMhdSwoeeEWYYrVB1IPVsQsrn/Pg9fpCOLp0O6A3gay4O0BiUAKCg4K7rvoD1osg/WOiUN/cNw+RIoQ49BjZ6QIBHEgPFc+6vygM//C8zh9u45QuARV+V/G7CUX+ouIEbZ/JqQOP2H88z3rIJlx0LRdae8Xx851eRM7Z1Biaiwv5DQWKPW4Gs1N0YkRPXomxq9Im6xt2BQPNrqe4PsbgUqZ5tYFPP8iHyHza8RAifLZ0D6FHwNqyZLIF92M4hcZoPA7zO+BXuyQ0IK4RoAdXAW4EY9QWr6msT0QCKcEVsuQzXNY0h5mDPVLGghcDhz+sfkR3WHsFqo3oyfJjD7hX9nMOOq6khf2PHZbo3+uCfTdpU7LzPH4kAFXD6fotG1pV8/y2tpkpXVMdZYoa9uuL9FBHgMVv5K8nH4ZFvgCFQAsly3S+ka1GO9ZY24S39HHxGfgzIAeUBS42bMKhzazWgfkMqf/KwsoeFdJHxVnS3LxAJ5HWCso=\n-----END CERTIFICATE-----"
```
Comment (jcrussell) [inline src/protonuke/ftp.go]: Should we use something akin to `generateCerts`?

**Comment by csymonds** [comment]
@jcrussell Applied your suggestions as well as some additional changes the make everything much more stable. Tested 100 clients hitting a single server at full speed for ~ 5 minutes and no issues.

**Comment by floren** [comment]
> jfflore@freyr:~/work/minimega$ sudo ./bin/protonuke -ftps -serve -level debug                  2017/03/01 16:11:54 DEBUG hosts.go:53: parseHosts                                    
2017/03/01 16:11:54 DEBUG protonuke.go:110: hosts: map[]                            
2017/03/01 16:11:54 DEBUG protonuke.go:116: enabling reports every 10s              
2017/03/01 16:11:54 DEBUG ftp.go:261: Image served will be 357 by 357   
2017/03/01 16:11:55 DEBUG generate_cert.go:86: wrote cert to: /tmp/protonuke_cert_256772515    2017/03/01 16:11:55 DEBUG generate_cert.go:94: wrote key to: /tmp/protonuke_key_872908966
2017/03/01 16:11:55 DEBUG logger.go:20:    protoFTP listening on 21                            2017/03/01 16:11:56 DEBUG logger.go:20: f5484f785032b83dc677   Connection Established
2017/03/01 16:11:56 DEBUG logger.go:36: f5484f785032b83dc677 < 220 Welcome to the Go FTP Server
2017/03/01 16:11:57 DEBUG logger.go:20: f5484f785032b83dc677   Connection Terminated           
2017/03/01 16:12:02 DEBUG logger.go:20: 991872c0be9d1079cb1a   Connection Established
2017/03/01 16:12:02 DEBUG logger.go:36: 991872c0be9d1079cb1a < 220 Welcome to the Go FTP Server
2017/03/01 16:12:03 DEBUG logger.go:31: 991872c0be9d1079cb1a > USER anonymous
2017/03/01 16:12:03 DEBUG logger.go:36: 991872c0be9d1079cb1a < 331 User name ok, password requi
red                                                                                     
2017/03/01 16:12:03 DEBUG logger.go:29: 991872c0be9d1079cb1a > PASS ****
2017/03/01 16:12:03 DEBUG logger.go:36: 991872c0be9d1079cb1a < 230 Password ok, continue
2017/03/01 16:12:04 DEBUG logger.go:31: 991872c0be9d1079cb1a > SIZE /tmp/ftpimage              2017/03/01 16:12:04 Size: error(lstat /home/john/work/minimega/tmp/ftpimage: no such file or di
rectory)                                                             
2017/03/01 16:12:04 DEBUG logger.go:36: 991872c0be9d1079cb1a < 450 path /tmp/ftpimage not found

This surely isn't right. Same thing happens with ftp and ftps

**Comment by csymonds** [comment]
@floren There was an issue with how the server package was resolving the specified directory. For now, I am just serving info from memory. Later I want to go back and allow the user to specify a directory to serve from, but for now this should fix the issue. Thanks.

**Comment by mkunz7** [comment]
```
root@ubuntu:~/test3/minimega# ./all.bash
CHECKING SOURCE CODE
./protonuke/ftpdriver.go
diff ./protonuke/ftpdriver.go gofmt/./protonuke/ftpdriver.go
--- /tmp/gofmt166127773 2017-08-29 14:24:19.709417735 -0700
+++ /tmp/gofmt800339288 2017-08-29 14:24:19.709417735 -0700
@@ -91,7 +91,7 @@
                owner: "protonuke",
                group: "protonuke",
                size:  int64(len(FTPImage)),
-               dir: true,
+               dir:   true,
        }

        return mockFileInfo, nil
gofmt - FAILED
protonuke/ftpdriver.go:162: unreachable code
exit status 1
go vet - FAILED

```
plz fix :)

**Pull request merged**